File: src/main/java_backup/com/example/liferay/reactpoc/ssr/SsrRenderer.java
package com.example.liferay.reactpoc.ssr;

import org.osgi.service.component.annotations.Component;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import java.io.InputStreamReader;
import java.io.Reader;

@Component(immediate = true, service = SsrRenderer.class)
public class SsrRenderer {

    private ThreadLocal<ScriptEngine> engineHolder = ThreadLocal.withInitial(() -> {
        ScriptEngine nashorn = new ScriptEngineManager(null).getEngineByName("nashorn");
        try {
            nashorn.eval(read("/META-INF/resources/dist/server.js"));
        } catch (ScriptException e) {
            throw new RuntimeException(e);
        }
        return nashorn;
    });

    public String render() {
        try {

            Object html =
                ((Invocable) engineHolder.get()).invokeFunction("renderApp");
            return String.valueOf(html);
        } catch (Exception e) {
            throw new IllegalStateException("failed to render in JavaScript", e);
        }
    }

    private Reader read(String path) {
        return new InputStreamReader(getClass().getClassLoader().getResourceAsStream(path));
    }

}

File: src/main/kotlin/com/example/liferay/reactpoc/ssr/SsrRenderer.kt
package com.example.liferay.reactpoc.ssr

import org.osgi.service.component.annotations.Component
import java.io.InputStreamReader
import java.io.Reader
import javax.script.Invocable
import javax.script.ScriptEngineManager

@Component(immediate = true, service = arrayOf(SsrRenderer::class))
class SsrRenderer {

    private val engineHolder = ThreadLocal.withInitial {
        val nashorn = ScriptEngineManager(null).getEngineByName("nashorn")
        nashorn.eval(read("/META-INF/resources/dist/server.js"))
        nashorn
    }

    fun render(): String {
        try {
            val html = (engineHolder.get() as Invocable).invokeFunction("renderApp")
            return html.toString()
        } catch (e: Exception) {
            throw IllegalStateException("failed to render in JavaScript", e)
        }

    }

    private fun read(path: String): Reader {
        return InputStreamReader(javaClass.classLoader.getResourceAsStream(path))
    }

}